bsip-0040.md

```diff
@@ -0,0 +1,322 @@
+and if all passes continue with statefull asserts. If `now > interval_began + interval_in_sec`, then set `max_cumsum = 0` and set `interval_began = now`.
+The assert that needs to pass is now `current_cumsum + incoming value <= max_cumsum`. If all `asserts` of this `custom_active_authority` pass, update `current_cumsum = current_cumsum + incoming value`.
+
+#### `limit_monthly`
+Statefull assert, only `int` type arguments. Analogue to `limit`, but `interval_began` is initially set to `month(valid_from)` and set to `month(now)` on update, additionally the time assert is `month(now) >= interval_began + interval_in_months` (include logic for month overflow when year changes).
+
+#### `attribute_assert`
+Stateless assert, only for dictionary type objects. The data list contains restrictions that all must pass, the reference for the `argument` of the child restriction is nested into the attributes of the parent dictionary type object. Allows nesting of `attribute_assert`.
+
+#### `logical_or`
+Stateless assert, only for dictionary type objects. The data is a list of restrictions lists, i.e. 
+`data = [ [restriction 1, restriction 2], [restriction 3, restriction 4], ... ]`. If one of the restrictions sub-lists in data passes as whole, this restriction passes. In above miniexample that would mean if `restriction 1` and `restriction 2` pass, the whole `logical_or` restriction is considered to be passed as well.
+
+### Outline of handling incoming transactions
+ 
+When a signed transaction arrives and before the backend evaluates if all necessary authorities are present through the signatures, do the following:
+- iterate over required accounts and for each account, iterate over all operations within the transactions that require the active authority of this account
+- iterate the `custom_active_authorities` of said account
+- if a `custom_active_authority` is found that matches , remember that and stop iterating the authorities and continue until all operations are checked
+- if the account has a `custom active authority` match for every operation in the transaction that requires it, then grant the `active authority` of said account. If no match is found, treat as if no authority was given
+
+Note:
+- A `custom_active_authority` can only grant the `active authority` of the corresponding account, nothing more
+
+### Economics
+
+Adding a custom active authority means increased effort for the backend, and with a stateful one also the need for more storage. Proposed transaction fees:
+- `install_custom_active_authority`: Tied to the duration of the custom active authority. 
+Normal accounts can only create custom active authoritites with a duration of maximum 1 year. LTM can do any duration and also unlimited, but the transaction fee is capped at duration of 2 years.
```

Comment:
Too complicated IMO. Better use fixed fee and duration limit (duration limit configurable through committee?).

Reply:
Hm. Idea is to have the user really think about how long they want it, as it increases RAM and computational effort server side.

I also added now that expired custom_active_authorities should be deleted on maintenance.

Reply:
How about `fee = flat_fee + basic_fee * duration`?
For update, `fee = flat_fee + max(0, delta(basic_fee * duration))`.
For removal, `fee = flat_fee`.

Reply:
>I also added now that expired custom_active_authorities should be deleted on maintenance.

I'd recommend disable it rather than removing it, since it's easier for users to "renew" an existing custom authority than to reinstall a new one. We can remove expired authorities after a period, for example a month or 7 days, to save RAM.

Reply:
> I'd recommend disable it rather than removing it, since it's easier for users to "renew" an  existing custom authority than to reinstall a new one. We can remove expired authorities after a period, for example a month or 7 days, to save RAM.

Good case, included as well.

Reply:
> How about fee = flat_fee + basic_fee * duration?
> For update, fee = flat_fee + max(0, delta(basic_fee * duration)).
> For removal, fee = flat_fee.
> 
> Calculate basic_fee according to complexity, e.g. size of authority, number of restrictions and etc.

Personally, I like it. Not fixated on how this is implemented. Have included it for now, if anyone has other input?